### PR TITLE
chore: support external types in typescript codegen

### LINKED ISSUE #9623: Typescript API Generator to generate OptionSet

### PR DESCRIPTION
**Pull request by Emyrk**
Closes https://github.com/coder/coder/issues/9623

We now have a lot of structs in the `clibase` package that need supporting.

Unfortunately, some of this code is very "bolt on", and not perfectly clean. A problem is that we override the `MarshalJSON` function on some types, so those need to be handled case by case.

The outputted ts file is the test imo.

### CODE REVIEW

File: site/src/api/typesGenerated.ts [hunk 1]
```diff
@@ -2071,3 +2042,48 @@ export interface DerphealthStunReport {
...
+
+// From clibase/option.go
+export type ClibaseValueSource = "" | "default" | "env" | "flag" | "yaml";
+export const ClibaseValueSources: ClibaseValueSource[] = [
+  "",
+  "default",
+  "env",
+  "flag",
+  "yaml",
+];
```
Comment (Emyrk) [inline site/src/api/typesGenerated.ts:2089]: @BrunoQuaresma

**Review by mafredri — APPROVED** [review]
> Unfortunately, some of this code is very "bolt on", and not perfectly clean.

I think that's perfectly fine for this package! Don't let perfect be the enemy of good, and all that.

LGTM!

**Comment by Emyrk** [comment]
> > Unfortunately, some of this code is very "bolt on", and not perfectly clean.
> 
> I think that's perfectly fine for this package! Don't let perfect be the enemy of good, and all that.
> 
> LGTM!

Totally agree, just unfortunate. Originally we thought we could open source this as a stand alone package. So every "hard coded" opinion makes that more challenging. It's obviously not a high priority, would just be cool :shrug:

**Comment by Emyrk** [comment]
That most challenging part about making it not opinionated is the override of `MarshalJSON`. There really is no good way to figure out what the typescript type is from that without some much more advanced functionality.

**Comment by mafredri** [comment]
> That most challenging part about making it not opinionated is the override of `MarshalJSON`. There really is no good way to figure out what the typescript type is from that without some much more advanced functionality.

Yeah that’s true, but perhaps it’s not something worth solving by analyzing the custom marshaler. Maybe it could be solved by a separate interface that you can implement which, when implemented, returns a value (concrete type) that “looks” correct/contains all the possible values in the case of an enum. This could work with custom types within a struct as well if those too implement it.

Not sure if it’s doable, just a thought I had.

**Comment by Emyrk** [comment]
@mafredri I was having a similar thought of having the user provide the marshalled struct via some other means. Because the only way I could think of having the custom marshaller analyzed without too much hassle, was to just run the marshal with some fuzzed data and see what pops out lol. But that seems flakey and probably not perfect.

**Pull request merged**